File: World Finals 2008/B.rb
#!/usr/bin/ruby
#
# Problem: Ping Pong Balls
# Language: Ruby
# Author: KirarinSnow
# Usage: ruby thisfile.rb <input.in >output.out
# Comments: Times out on large.


def txc(x, y)
  return $x0 + x*$vx1 + y*$vx2
end
  
def tyc(x, y)
  return $y0 + x*$vy1 + y*$vy2
end

def inbounds(x, y)
  tx = txc(x, y)
  ty = tyc(x, y)
  if (tx < 0) or (tx >= $w) or (ty < 0) or (ty >= $h) then
    return false
  end
  return true
end
   
def compute()
  $w, $h = readline.split.map(&:to_i)
  $vx1, $vy1 = readline.split.map(&:to_i)
  $vx2, $vy2 = readline.split.map(&:to_i)
  $x0, $y0 = readline.split.map(&:to_i)
 
  if $vx1*$vy2 - $vx2*$vy1 == 0 then
    v = Hash.new
    q = [[$x0, $y0]]

    while q.length > 0
      x, y = q.pop
      if v[[x, y]] == nil then
        v[[x, y]] = 1
        if (x+$vx1 >= 0) and (x+$vx1 < $w) and
            (y+$vy1 >= 0) and (y+$vy1 < $h) then
          q.push([x+$vx1, y+$vy1])
        end
        if (x+$vx2 >= 0) and (x+$vx2 < $w) and
            (y+$vy2 >= 0) and (y+$vy2 < $h) then
          q.push([x+$vx2, y+$vy2])
        end
      end
    end
    return v.length
    
  else
    j = 0
    l = 0
    u = 0
    c = 0
    while inbounds(j, u+1)
      u += 1
    end
    c += (u-l)+1

    j += 1
    while true
      while not inbounds(j, l)
        l += 1
        if l > u
          return c
        end
      end

      if inbounds(j, u) then
        while inbounds(j, u)
          u += 1
        end
        u -= 1
      else
        while not inbounds(j, u)
          u -= 1
        end
      end

      c += u-l+1
      j += 1
    end
  end
end

for $i in 1 .. gets.to_i:
    print "Case \##$i: %s\n" % compute()
end

File: World Finals 2008/B.cpp
// Problem: Ping Pong Balls
// Language: C++
// Author: KirarinSnow
// Usage: g++ thisfile.cpp -o executable && ./executable <input.in >output.out


#include <stdio.h>
#include <iostream>
#include <set>
#include <vector>

using namespace std;


long long int w, h, vx1, vy1, vx2, vy2, x0, y0;

bool inbounds(long long int x, long long int y)
{
    long long int tx = x0 + x*vx1 + y*vx2;
    long long int ty = y0 + x*vy1 + y*vy2;
    if ((tx < 0) || (tx >= w) || (ty < 0) || (ty >= h))
	return false;
    return true;
}

long long int compute()
{
    cin >> w >> h >> vx1 >> vy1 >> vx2 >> vy2 >> x0 >> y0;
    
    if (vx1*vy2 == vx2*vy1)
    {
	long long int c = 0;
	set<pair<long long int, long long int> > v;
	vector<pair<long long int, long long int> > q;
	q.push_back(make_pair(x0, y0));

	while (q.size() > 0)
	{
	    pair<long long int, long long int> p = q.back();
	    q.pop_back();
	    long long int x = p.first;
	    long long int y = p.second;
	    if (v.find(make_pair(x, y)) == v.end())
	    {
		v.insert(make_pair(x, y));
		if ((x+vx1 >= 0) && (x+vx1 < w) &&
		    (y+vy1 >= 0) && (y+vy1 < h))
		    q.push_back(make_pair(x+vx1, y+vy1));
		if ((x+vx2 >= 0) && (x+vx2 < w) &&
		    (y+vy2 >= 0) && (y+vy2 < h))
		    q.push_back(make_pair(x+vx2, y+vy2));
	    }
	}
	return v.size();
    }
    else
    {
	long long int j, l, u, c;
	j = l = u = c = 0;
	while (inbounds(j, u+1))
	    u++;
	c += (u-l)+1;

	for(long long int j = 1; ; j++)
	{
	    while (!inbounds(j, l))
	    {
		l++;
		if (l > u)
		    return c;
	    }
	    if (inbounds(j,u))
		while (inbounds(j, u+1))
		    u++;
	    else
		while (!inbounds(j, u))
		    u--;
	    c += u-l+1;
	}	   
    }
}

int main()
{
    int cases;
    cin >> cases;
    
    for (int i = 1; i <= cases; i++)
    {
	printf("Case #%d: %lld\n", i, compute());
    }
    
    return 0;
}
